### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by IvanMalison** [travitch/taffybar#217]
#216 seems to have finally made taffybar stable enough for me that it has become clear that there is a massive memory leak in the program. Memory usage crept up to 70% on my 16 gig machine in about 24 hours. I have no suspicions about the cause yet.

**Comment by simlu** [comment]
In that case the crash and auto restart is actually preventing me from having any major problems with this (no major pun intended). Two bugs helping each other out...

**Comment by IvanMalison** [comment]
So no one lese has ever gotten this:

Hopefully this is not something introduced by #237 . I just got massive memory usage from taffybar today:
https://imgur.com/f5GmErO

**Comment by IvanMalison** [comment]
Wow, this has gotten bad, #229 #237 have made things stable enough to make this the biggest/only source of crashes now...

**Comment by sakshamsharma** [comment]
Memory leak in progress:
https://imgur.com/fG8TxfE

**Comment by sakshamsharma** [comment]
I hadn't used taffybar in the past month or so. Today, I have noticed very speedy memory usage rise.

I played around with profiling a bit, but since stack rebuilds itself after reading config, I will have to make it rebuild itself, so that it can profile the final binary too.

**Comment by IvanMalison** [comment]
@sakshamsharma yeah its kind of a pain. I think you basically need to skip the dyre stuff and manually recompile to do any profiling.

**Comment by sakshamsharma** [comment]
Someone will have to do that at some point.

Anyhow, I was able to reproduce excessive memory usage patterns by moving around workspaces, opening terminals and closing them, such acts as fast as I could, for around 2 minutes.

**Comment by IvanMalison** [comment]
@sakshamsharma 

oh thats good to hear. I haven't been able to consistently reproduce the issue. If you could write a script that does it that would be amazing!

**Comment by IvanMalison** [comment]
@sakshamsharma 
Do you use any custom icons? I've sort of always suspected that this might be related to icon usage, and specifically #243 somehow.

**Comment by sakshamsharma** [comment]
@IvanMalison I do use icons, but they're all installed at the system level. I had manually installed the emacs icon somewhere too (It is the same as your display picture on Github :smile: ). I don't understand what "custom icons" means here.

Let's discuss this further on gitter?

**Comment by weskerfoot** [comment]
I'm also seeing massive memory usage to the point that my system was close to invoking the OOM killer.

**Comment by IvanMalison** [comment]
tray =/ custom icons. Custom icons are for the display of icons for windows in workspacehud.
I no longer think that custom icons are the issue anyway. I do think it must be related to the FFI somehow.

It might be time to use valgrind to evaluate what is going on.

**Comment by LeifW** [comment]
On passing flags for profiling - you can add GHC options in`Dyre.ghcOpts` here:
https://github.com/travitch/taffybar/blob/master/src/System/Taffybar.hs#L301
and the executable Dyre produces is at `$HOME/.cache/taffybar/taffybar-linux-x86_64` - you can just execute that directly to get your main and not the Dyre main.

**Comment by IvanMalison** [comment]
@LeifW Yeah, but unfortunately ghc profiling is not much help if the issue is in C code. I did do a profile run sometime last year, and I vaguely remember getting pretty useless results, which again, points to gtk/X11 being the source of the issue.

We should probably profile again just to make sure though.

**Comment by PierreR** [comment]
I am also experiencing what I believe is abnormal mem usage after upgrading to taffybar-1.0.1:

![taffybar_mem](https://user-images.githubusercontent.com/93115/38853470-9ea65652-421d-11e8-8770-dc4f2d4f5a1e.png)
 
I also have had a freezing of the whole xmonad which is quite unusual.

**Comment by IvanMalison** [comment]
the freezing is probably just due to OOM from taffybar

**Comment by weskerfoot** [comment]
@IvanMalison @simlu 
I ran taffybar under strace and reproduced the memory leak (at least the one that I was getting).
I've uploaded the log here https://gist.github.com/weskerfoot/5dcadd3315d3f49f5b545e11e626920f
It seems the majority of the system calls made when the memory usage increases are things like
```
futex(0x11a7b88, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x2ecae6c, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x2ecae70, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0x7f421c00da78, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x7f421c00da80, FUTEX_WAKE_PRIVATE, 1) = 1
```
I didn't know what a futex was, but apparently it has something to do with asynchronous locking. Could the issue have something to do with the RTS?
http://man7.org/linux/man-pages/man2/futex.2.html
I also noticed that whenever I switched windows it seemed to bump the memory usage up quite a bit.

**Comment by weskerfoot** [comment]
There's also some info here that might be potentially relevant to gtk-traymanager: https://ghc.haskell.org/trac/ghc/wiki/Commentary/Rts/Signals
> When using foreign libraries through the Haskell FFI, it is important to ensure that the foreign code is capable of dealing with system call interrupts due to alarm signals GHC is generating.

**Comment by IvanMalison** [comment]
> I didn't know what a futex was, but apparently it has something to do with synchronization. Could the issue have something to do with the RTS?
> http://man7.org/linux/man-pages/man2/futex.2.html
> I also noticed that whenever I switched windows it seemed to bump the memory usage up quite a bit.

If there are more wait than wakes, that could indicate that some type of deadlock on some shared resouce, and the memory leak is due to things accumulating that are waiting for the share resource.

**Comment by nyarly** [comment]
You might also try out memleak at
https://github.com/iovisor/bcc/blob/master/tools/memleak.py Extremely
handy, I've found.

**Comment by weskerfoot** [comment]
https://gist.github.com/weskerfoot/5dcadd3315d3f49f5b545e11e626920f#file-strace-out-L2688
This seems like it could be an issue. It's not finding the gdk pixbuf message catalog file. If I create it:
```
sudo cp /usr/share/locale/en_CA/LC_MESSAGES/gdk-pixbuf.mo /usr/share/locale/en_US.UTF-8/LC_MESSAGES/
```
And then run strace again, then I can see that it's able to successfully open that file. I'll monitor memory usage again to see if that has an effect.

Anyone know specifically what that file is used for?

**Comment by IvanMalison** [comment]
@weskerfoot @nyarly So you guys are able to reproduce consistently?

I still haven't figured out how to force the memory leak to happen.

**Comment by nyarly** [comment]
My reproduce was "let it run for 4 hours."

**Comment by weskerfoot** [comment]
@IvanMalison for me it happens if I open Spotify, then open two windows and start alt-tabbing between them. Every time I switch windows it uses a bit more memory. It only happens when Spotify is running for some reason.

**Comment by IvanMalison** [comment]
Yeah, so I don't think that the issue is necessarily with spotify, because memory usage still increases without it, its just that having spotify open makes the problem super obvious because of how big the icon is in the case of spotify, but this is a pretty clear indication that the problem is 100% occuring with the icon images stuff.

**Comment by IvanMalison** [comment]
@weskerfoot @nyarly @simlu @sakshamsharma @PierreR @LeifW I assume that all of you are using IconImages in some for or another, correct?

The two spots that I currently suspect are

https://github.com/travitch/taffybar/blob/b57f81775bb5d574f37f886226eae1a7c53e6806/src/System/Taffybar/IconImages.hs#L77

(We never free this memory)

and

https://github.com/travitch/taffybar/blob/b57f81775bb5d574f37f886226eae1a7c53e6806/src/System/Information/SafeX11.hsc#L61

Although this should be safe because of

https://github.com/travitch/taffybar/blob/b57f81775bb5d574f37f886226eae1a7c53e6806/src/System/Information/SafeX11.hsc#L93

So I think that first one is our best bet.

**Comment by IvanMalison** [comment]
@weskerfoot @nyarly @simlu @sakshamsharma @PierreR @LeifW 
Fix is ready in #278 .

I haven't merged to master yet because this change requires a containerized build (#272) because travis still runs trusty which has an ancient version of gtk3.

The fix is also merged in to https://github.com/travitch/taffybar/tree/use_gtk-strut which will actually probable get merged in to master after the containerized build issue is fixed anyway. That branch is really what I'd like people to use since it has pretty significant interface changes.

**Comment by weskerfoot** [comment]
@IvanMalison thanks for fixing that. I'll be sure to pull today and test it out.

**Comment by IvanMalison** [comment]
@weskerfoot Just a heads up. There are some pretty drastic (backwards incompatible) changes on master. I'd love it if you would also test those, but totally understand that not everyone will want to, which is why I released 1.0.2, which has the memory leak fix, but none of the other stuff that is on master.

**Comment by GregorySchwartz** [comment]
I am seeing a severe memory leak in version 2.1.1 from hackage.

**Comment by simlu** [comment]
^ Auto kill and restart on timer has been very reliable here. As a workaround :)

**Comment by IvanMalison** [comment]
@simlu You also still see the memory leak in the newest version?

**Comment by IvanMalison** [comment]
@GregorySchwartz @simlu The issue causing the memory leak described in this issue is well understood, and definitely fixed.

If you guys have a memory leak, you should file a new issue.

**Comment by simlu** [comment]
Sounds good. I'll dig into it when I have some more time (I've heavily customized my config, so who knows what's causing it). But restarting on a five minute timer works really well as hacky as it sounds.

**Comment by simlu** [comment]
Yes. I'm suspecting it might have to do with my config, but didn't have time to deep dive yet. I'll open a new issue when I get to it.

**Comment by IvanMalison** [comment]
@GregorySchwartz @simlu  ..although, the following just occured to me...although, the following just occured to me.

You may need to be using haskell-gi of version at least 0.21.2


here is the relevant portion of the chat from gitter:
```

Ivan Malison @IvanMalison Apr 18 21:18
Is it correct that pixbufNewFromData frees its data_ argument?
https://hackage.haskell.org/package/gi-gdkpixbuf-2.0.15/docs/src/GI-GdkPixbuf-Objects-Pixbuf.html#pixbufNewFromData
It seems to me the argument should not be freed, because it will be used (in place) by the pixbuf.
Which is why the PixbufDestoryNotify is offered (so that the data can be freed when the pixbuf is destroyed)

Iñaki @garetxe Apr 19 08:56
@dylex Which signal do you have in mind? Right now there is support for notify::prop-name, so you can do
on widget (PropertyNotify #propName) $ do ...
Not for other signals, but it should be simple enough to add support for something like
on' widget #signal "detail" $ do ...

Iñaki @garetxe Apr 19 09:02
@IvanMalison No, you are completely right, it should not free data_. That's a bug in haskell-gi! I will fix this asap, but could you please file a bug so it does not get lost? Thanks!

Iñaki @garetxe Apr 19 12:41
@IvanMalison This should now be fixed by haskell-gi/haskell-gi@b603ff8. Thanks again for the report!
I will release a new version of haskell-gi with the fix soon.

Iñaki @garetxe Apr 19 12:56
(Looking to pixbufNewFromData I noticed that there is a small leak too, I'll fix that before releasing too.)

Ivan Malison @IvanMalison Apr 19 13:49
@garetxe What is the leak out of curiosity?

Iñaki @garetxe Apr 20 02:32
@IvanMalison Notice that if destroyFn is Nothing then ptrdestroyFn will leak. Of course, this is only sizeof(void*) bytes (so 8, typically), that will leak only if data_ leaks, but it needs to be fixed.

Ivan Malison @IvanMalison Apr 20 10:09
right. Do you think you will release soon? I'd like to pixbufFromData in a project.

Iñaki @garetxe Apr 20 11:17
Done, haskell-gi-0.21.2 should have the fix. 
```

@simlu @GregorySchwartz which install approach are you using? If you are using the aur package, I really recommend using stack, cabal or nix to install taffybar. See https://github.com/taffybar/taffybar/issues/151#issuecomment-360981068 for more details.

**Comment by LeifW** [comment]
I don't understand the aversion to the AUR package. It is built with cabal, same as you would get by using cabal, stack, or nix.

**Comment by IvanMalison** [comment]
@LeifW It has to do with the fact that it uses dynamic linking, and that it requires adding all dependencies separately from the aur.

See https://wiki.archlinux.org/index.php/Haskell#Problems_with_linking

One crucial part is:

> Since GHC provides no ABI compatibility between compiler releases, static linking is often the preferred option for local development outside of the package system.

It also means that installed packages don't respect version number dependencies:

https://www.reddit.com/r/archlinux/comments/7hanz7/whats_the_current_state_of_haskell_on_arch/dqprktz/

I have come to terms with the fact that many people seem to really strongly prefer installation from the AUR, so I have no problem with having it as an option (Though I do think that it would be better to have the aur package simply set up as stack install with stack static). My concern is mainly that I think it can lead to a subpar experience, and issues that have to do with some of the issues that I mentioned above.

**Comment by IvanMalison** [comment]
> It is built with cabal, same as you would get by using cabal, stack, or nix.

There is a very meaningful sense all of these methods are not the same. Specifically, they all differ w.r.t how they pull in dependencies. Installing with the aur loads the dependencies from the aur, cabal from hackage, stack from stackage, and nix from nixpkgs.

The aur's versioning system leaves a lot to be desired (basically doesn't exist), and because of the lack of ABI compatibility, you can very easily install versions of cabal packages that won't work when e.g. ghc is upgraded. This ultimately (though perhaps not initially) introduces additional cognitive load for users.

I feel like the reason so many people want an AUR package is because they don't want to have to think about haskell, and especially its packaging system, too much, but the reality is that the solution it offers is almost certain to lead to problems down the line, and require them to understand MORE about haskell packaging than they would have if they had just used one of the existing tools.

@LeifW Do you think all of these concerns are illegitimate? You seem to have a fairly good understanding of how haskell works -- why do you prefer an AUR installation?

**Comment by weskerfoot** [comment]
Just jumping in here, I don't feel like an AUR version is going to be useful unless it builds it using ghc-static or something, in which case, it's probably easier to just install it using this method https://wiki.archlinux.org/index.php/Haskell#Building_statically_linked_packages_with_Cabal_.28without_using_shared_libraries.29

Personally, I also have this in my `/etc/pacman.conf` because I've also had issues with GHC upgrades breaking XMonad
```
IgnorePkg = ghc
IgnorePkg = ghc-static
IgnorePkg = ghc-libs
```

An alternative would be to install it using Nix, but I wasn't able to get that working easily either (can't remember exactly what the issue I had was), but installing GHC and Cabal with Nix would be ideal.

**Comment by IvanMalison** [comment]
> it's probably easier to just install it using this method https://wiki.archlinux.org/index.php/Haskell#Building_statically_linked_packages_with_Cabal_.28without_using_shared_libraries.29

Yes, this is what I have sort of always maintained. It might be sort of nice to try to write an aur package to automate all of this, but it could be kind of tricky because cabal and ghc-pkg are so stateful.

Not sure if you are proposing to use:

**Comment by weskerfoot** [comment]
No I'm not proposing using Nix to install via PKGBUILD or anything, I don't think that would be a good idea. I was proposing maybe people who don't want to use cabal or stack could just use nix.

**Comment by LeifW** [comment]
This is how I've been installing Arch packages for a decade. The Haskell packages in AUR were formerly from [Don Stewart](https://archhaskell.wordpress.com/2009/07/25/arch-haskell-news-july-25-2009/), generated with his cabal2arch tool. I like managing everything with pacman (get upgrades, query files, uninstall, etc) instead of using a tool per language and managing things installed into my home dir. Especially for things that depend on C libraries on install and track.
Also, if packages on AUR become popular enough, they can be promoted into [community]. Thanks to that, people can now simply `pacman -S idris` or agda or whatever without figuring out how to install it. We could also run our own binary repo of pre-built packages to offer.

>The aur's versioning system leaves a lot to be desired (basically doesn't exist), and because of the  >lack of ABI compatibility, you can very easily install versions of cabal packages that won't work when >e.g. ghc is upgraded. This ultimately (though perhaps not initially) introduces additional cognitive load >for users.

I'm unclear what this is referring to. The fact that the checksum of the library build isn't put into the version number? I don't really see what the "danger" is. An compile + install is a compile + install, regardless of method. No Haskell library will work after you have removed the GHC version it was built against. If you want to avoid that, you can have multiple GHC versions on your system (a la Nix, Stack, or just installing to different paths).

Static vs dynamic linking isn't particularly relevant for libraries (they need to be present at compile time either way), and taffybar isn't solely an executable.
GHC has supported dynamic linking for nearly a decade, and I actually prefer it that way. Saves me plenty of disk space and compile time.

The downside is that when Haskell packages I'm using in [community] get upgraded, the Haskell things I've installed break. So when I see haskell packages scroll by during a `pacman -Syu`, afterwards I run `ghc-pkg check` to see what was broken, and reinstall any libs reported. Then I run `xmonad --recompile` and re-run taffybar to force it to recompile. That's it. Do you have other concerns? To me, that trade-off is worth it. I use Arch *because* things get upgraded and kept current.
I'd rather not install a whole separate GHC and set of libraries to manage myself, when the xmonad provided by community works fine for me. And since my xmonad.hs depends on both xmonad and taffybar, for sanity's sake, I think it's best to use the same method to install both.
Some pros and cons of different methods are mention on the Arch Wiki: https://wiki.archlinux.org/index.php/Haskell#Pros.2FCons_of_the_different_methods

> There is a very meaningful sense all of these methods are not the same. Specifically, they all differ w.r.t how they pull in dependencies. Installing with the aur loads the dependencies from the aur, cabal from hackage, stack from stackage, and nix from nixpkgs.

In all cases tarballs of the packages are downloaded from hackage, and cabal builds them with GHC.
The details are relevant in installation and upgrading. Not really for runtime issues, such as memory leaks. GHC and/or package versions may be relevant, though.

**Comment by LeifW** [comment]
You could make a PKGBUILD that installs taffybar via stack. Though the library needs to be available to the same GHC you're using for xmonad. Since it's a library it doesn't really matter if you use static or dynamic for that - dynamic just saves some disk space and compile time.
It might be useful to have a library that depends on all the C libs taffybar depends on, e.g. a "taffybar-pre-reqs" for people to install? Or just provide the pacman command that installs all those.

**Comment by IvanMalison** [comment]
> I'm unclear what this is referring to. The fact that the checksum of the library build isn't put into the version number? I don't really see what the "danger" is. An compile + install is a compile + install, regardless of method.

I'm referring to the fact that the aur makes no attempt to manage version numbers at all. If someone upgrades one of taffybar's dependencies (say gi-gtk) to a version that is incompatible with what is required by taffybar, the package will no longer be installable. This is an obstacle to maintenence, and it seems far less robust than being able to depend on a specific pinned version number.

Here is another case to consider. Say that a bug gets fixed in a haskell dependency of taffybar's that would still allow taffybar to compile -- There is no way to for me to enforce that taffybar will use an updated version of the dependency to fix the bug.

> No Haskell library will work after you have removed the GHC version it was built against.

True, but managing ghc installs with aur/pacman is the only way a ghc installation can be upgraded to a new version without an explicit request by the user. It's not super easy, but it is at least possible to manage multiple ghc installations with cabal. Also everything wont just break instantly if you install a new version of ghc for cabal.

> If you want to avoid that, you can have multiple GHC versions on your system (a la Nix, Stack, or just installing to different paths).

Exactly why stack is preferred. The build is just much more robust, the only downside, as you mentioned is that you cant specify the non-haskell dependencies, but of course, this could be ameliorated if we made an aur package that installed with stack.

> The downside is that when Haskell packages I'm using in [community] get upgraded, the Haskell things I've installed break. So when I see haskell packages scroll by during a pacman -Syu, afterwards I run ghc-pkg check to see what was broken, and reinstall any libs reported. Then I run xmonad --recompile and re-run taffybar to force it to recompile. That's it.

That's it?

Yes, this is 90% of the concern. It may seem insignificant to you, but I think that sort of user experience can be very confusing, especially for those who are unfamiliar with haskell (who are probably the ones who want to install with pacman/aur anyway).

> Do you have other concerns? To me, that trade-off is worth it. I use Arch because things get upgraded and kept current.

I'm not sure what you are getting at here, but things will always (and have historically) been more current on hackage than they are in the aur (unless you are installing from git I suppose). in the way of legitimate gripes:

> And since my xmonad.hs depends on both xmonad and taffybar

Only for pager hints right? The plan is to move that module in to xmonad-contrib so that taffybar no longer needs to house it.

Taffybar pulls in a pretty crazy amount of dependencies, and literally none of them are needed for xmonad at all, so you should probably avoid this IMO.

> I'd rather not install a whole separate GHC and set of libraries to manage myself, when the xmonad provided by community works fine for me. And then you have mentioned:

This makes it sounds like you have to actively manage things, when the reality is that you don't.

Provided that all non-haskell dependencies are installed, a stack install will just work, and if we made an aur stack package it would be much more reliable than whatIn all cases tarballs of the packages are downloaded from hackage, and cabal builds them with GHC we have.

The current approach will always involve messing with ghc-pkg and manual reinstalls.
in the way of legitimate gripes:
-------------------------------------------------
> Some pros and cons of different methods are mention on the Arch Wiki: https://wiki.archlinux.org/index.php/Haskell#Pros.2FCons_of_the_different_methods

For me, when I look at that table of cons under stack and I don't see much:

> Installed in home directory
Okay, a legitimate gripe, perhaps, but doesn't seem like a huge deal to me. Also very easy to compile a taffybar final binary that doesn't need compilation to install globally if that is really needed.

> versions are pinned to snapshot
You are always free to specify a specific version from hackage or git using extra deps.

The AUR approach suffers from the same issue except it is worse because there isn't an easy to describe way to use other versions of dependencies (short of explicitly messing with ghc-pkg).

I would actually argue that stack gives both the MOST flexibility AND stability here.

> difficult to remove specific packages
Everything is built/installed in its own sandbox why does this matter?

And then you mentioned that
> I'd rather not install a whole separate GHC and set of libraries to manage myself

Which makes it sound like you need to do something, which isn't true.

------------------------------------

> In all cases tarballs of the packages are downloaded from hackage.

Yes, but the crucial difference is that they all have different ways of selecting which tarball is downloaded from hackage.

> Not really for runtime issues, such as memory leaks. 

Dependency versions have an effect on runtime behavior. See the example that I gave above. This has been the case for several issues that were resolved recently.

-------------------------------------------------------

tl;dr:

Here are the issues I see with aur installation:
* There is no control of haskell dependency versions, which could make it so that this install approach fails at either runtime or install time. It also leads to the possibility of there being users out there running out of bounds sets of versions successfully which could lead to difficult to reproduce bug reports.
* Even when they are semantically compatible, dependency package updates can cause breakage of installed libraries because of ABI incompatibility. It is my understanding the reason this is particularly pronounced with aur managed packages because of dynamic linking, but you claim that this is not the case.
* aur installations of GHC can be upgraded at any time, borking existing haskell packages

In my experience, even as someone who is quite experienced with haskell, dealing with all of that is a huge pain, and very likely to:

* Turn off new users, who will actually be very likely to use this approach since they will view it as easier and avoiding the n-package managers problem
* lead to difficult to reproduce bug reports
* Involve extra maintenence work, even if only for the actual aur package itself

The first issue is by far the most important though. Having installation and updates not be robust is a huge deal breaker IMO.

**Comment by IvanMalison** [comment]
> You could make a PKGBUILD that installs taffybar via stack. Though the library needs to be available to the same GHC you're using for xmonad. Since it's a library it doesn't really matter if you use static or dynamic for that - dynamic just saves some disk space and compile time.

Not sure how familiar you are with stack, but this is not how it works. You can't make a package "available" to a particular ghc. Stack always installs and builds things in completely isolated environments, which is why it is necessary to add a .cabal and .stack file to the `$HOME/.config/taffybar` directory. Dyre currently has some limited stack support, so when it see the .stack file in the taffybar directory, it will actually execute the binary using stack exec rather than building with ghc etc.

It is because this setup is required that I have not yet made an aur package.

Nix has the best story for handling all of this, and is really going to be the only package manager that can actually have a single command build that works every time. Unfortunately, it simply does not have wide enough distribution to be a viable option here.

> It might be useful to have a library that depends on all the C libs taffybar depends on, e.g. a "taffybar-pre-reqs" for people to install? Or just provide the pacman command that installs all those.

yeah having those non-haskell dependencies as a separate package seems like a good idea.

**Comment by LeifW** [comment]
> I'm referring to the fact that the aur makes no attempt to manage version numbers at all. If someone upgrades one of taffybar's dependencies (say gi-gtk) to a version that is incompatible with what is required by taffybar, the package will no longer be installable. This is an obstacle to maintenence, and it seems far less robust than being able to depend on a specific pinned version number.
> 
> Here is another case to consider. Say that a bug gets fixed in a haskell dependency of taffybar's that would still allow taffybar to compile -- There is no way to for me to enforce that taffybar will use an updated version of the dependency to fix the bug.

You can absolutely specify version constrains on dependencies in a PKGBUILD - same as you would in a .cabal file. See https://wiki.archlinux.org/index.php/PKGBUILD#depends
This seems a little redundant to me, though, as the version constraints in the .cabal file are still enforced
When a haskell package is upgraded in Arch, they rebuild the packages that depend on it and increment  pkgrel field, because the package with the old version number is no longer valid. E.g. the version number for the [haskell-lens](https://www.archlinux.org/packages/community/x86_64/haskell-lens/) is currently 4.16.1-13. It's lens 4.16.1, and since the pkgrel is 13, that means that version has been rebuilt 12 times due to having the haskell libs it depends on upgraded.
You could also put the ghc build hash in the version number, too, if you wanted to pin them to specific builds.

And yes, breaking of AUR packages by swapping out their with ABI incompatible versions is an issue (the same as if you had cabal-installed libraries on top of Arch packages and then upgraded those packages) but dynamic linking doesn't cause any more breakages for libraries in that regard. The libraries will be just as broken with static or dynamic. It's only for executables where static linking will preserve your executable from arch library upgrades (because those libraries are already inlined into the executable).

We could include a disclaimer in the current taffybar PKGBUILD about alternate install methods / official support / etc. I've been happily using xmonad-extras out of AUR on top of xmonad-contrib out of [community] for a decade, but the less stuff in [community] a package leverages, the less sense it probably makes to build it on top of the [community] repo.

**Comment by LeifW** [comment]
@IvanMalison Understood. I think it'd be nice to ultimately see taffybar included in the [community] repo - then people could `pacman -S taffybar` the same as you can with xmonad, and get a taffybar supported by Arch without having to worry about the ABI breakages you get with building it yourself out of AUR and then having packages from [community] upgraded out from underneath it. People preferring to use stack or other tools would still likely want to stick with not installing any haskell libs out of [community] through.

**Comment by GregorySchwartz** [comment]
@IvanMalison I am using `stack install taffybar-2.1.1` with `haskell-gi-0.21.2`. @LeifW There are many users that do not use Arch Linux (I do, but that's besides the point). I prefer to build from Hackage or Stackage, that way I can have my entire system (using xmonad) on the same Stackage resolver. Especially because I have other Haskell packages in use for my taffybar configuration.

**Comment by IvanMalison** [comment]
@GregorySchwartz I just took a look at your taffybar config, and you don't seem to have a stack.yaml file or a cabal file. Are you sure that you are actually using the stack versions of the library with taffybar?

One of the subtleties/annoyances of using stack right now is that taffybar wont reexecute itself in the stack environment unless you explicitly set one up. If you have taffybar installed on your computer, it could be that you are running an older version of taffybar than the one that you have stack installed.

effectively, the only thing running `stack install taffybar-2.1.1` will do is install the taffybar binary from that version of the package.

**Comment by IvanMalison** [comment]
@GregorySchwartz It does seem however that you are using a sufficiently modern version of the taffybar config, and that therfore the memory leak issue should be fixed. Maybe you just havent checked in the stack.yaml and cabal files?

**Comment by IvanMalison** [comment]
@GregorySchwartz Also just a heads up that there is a pretty customizable and nice upower widget that you could rpobably use to replace your batString widget.

https://github.com/GregorySchwartz/dotfiles/blob/master/.config/taffybar/taffybar.hs#L134

**Comment by GregorySchwartz** [comment]
@IvanMalison I run taffybar with `stack exec taffybar` to make sure all of the correct versions of libraries are used. I should mention that I recently moved over to the newest version of taffybar, the one from maybe a year / year and a half ago was what I was using until now.

**Comment by GregorySchwartz** [comment]
Although I had the correct `haskell-gi` version in my `stack.yaml`, I copied the dependencies from this repository to my `stack.yaml` and reinstalled. It seems to be OK now, so I wonder if it's multiple dependencies, but it seems fine. Although it does seem to creep up over time (2% of memory to 2.6% over the course of hours), it was not nearly as aggressive as before (50% over the course of less than an hour).

**Comment by IvanMalison** [comment]
@GregorySchwartz Ahh yes, this is likely what it was. haskell-gi versions are actually pretty weird, because for things that are generated with haskell-gi like gdk-gdkpixbuf use the version of the dependency themselves and get re released under the same version. One of the fixes for the memory leak was to actually allow the freeing of pixbuf memory, so that is likely what you were coming up against.

**Comment by IvanMalison** [comment]
Also see #472
